src/Storages/MergeTree/RowOrderOptimizer.cpp

```diff
@@ -0,0 +1,162 @@
+#include <Storages/MergeTree/RowOrderOptimizer.h>
+
+#include <Interpreters/sortBlock.h>
+#include <base/sort.h>
+#include <Common/PODArray.h>
+#include <Common/iota.h>
+
+#include <numeric>
+
+namespace DB
+{
+
+namespace
+{
+
+/// Do the left and right row contain equal values in the sorting key columns (usually the primary key columns)
+bool haveEqualSortingKeyValues(const Block & block, const SortDescription & sort_description, size_t left_row, size_t right_row)
+{
+    for (const auto & sort_column : sort_description)
+    {
+        const String & sort_col = sort_column.column_name;
+        const IColumn & column = *block.getByName(sort_col).column;
+        if (column.compareAt(left_row, right_row, column, 1) != 0)
+            return false;
+    }
+    return true;
+}
+
+/// Returns the sorted indexes of all non-sorting-key columns.
+std::vector<size_t> getOtherColumnIndexes(const Block & block, const SortDescription & sort_description)
+{
+    const size_t sorting_key_columns_count = sort_description.size();
+    const size_t all_columns_count = block.columns();
+
+    std::vector<size_t> other_column_indexes;
+    other_column_indexes.reserve(all_columns_count - sorting_key_columns_count);
+
+    if (sorting_key_columns_count == 0)
+    {
+        other_column_indexes.resize(block.columns());
+        iota(other_column_indexes.begin(), other_column_indexes.end(), 0);
+    }
+    else
+    {
+        std::vector<size_t> sorted_column_indexes;
+        sorted_column_indexes.reserve(sorting_key_columns_count);
+        for (const SortColumnDescription & sort_column : sort_description)
+        {
+            size_t id = block.getPositionByName(sort_column.column_name);
+            sorted_column_indexes.emplace_back(id);
+        }
+        ::sort(sorted_column_indexes.begin(), sorted_column_indexes.end());
+
+        for (size_t i = 0; i < sorted_column_indexes.front(); ++i)
```

Comment:
The logic is too complicated. Can we not initialize `other_column_indexes` with all column indexes in the block, then iterate `sort_description` and remove the corresponding column indexes from `other_column_indexes`